### ISSUE TITLE
FORM element adjacent to a named field messes up overlays

### ISSUE THREAD

**Issue by lucasvreis** [minad/tempel#137]
It seems that having a FORM element right next to a field element can leave the overlays in a bad state. Consider the following two tempel expansions:
```elisp
(tempel-insert '((if (length> str 1) "{" "") " " (s str) " " (if (length> str 1) "}" "")))
(tempel-insert '((if (length> str 1) "{" "")     (s str)     (if (length> str 1) "}" "")))
(tempel-insert '((if (length> str 1) "y" "n")    (s str)     (if (length> str 1) "y" "n")))
```
In the first one, the FORM elements work as expected. In the second one, both FORM overlays appear behind the field, and in the third one both FORM overlays appear after the named field. They also stop updating correctly on further input.

**Comment by minad** [comment]
Yes, adjacent overlays (fields or forms) are not supported well. There have been reports about these problems before. Patches improving this are welcome, if possible. The problem is that we have no underlying text to pin the different overlays to, if they appear at the same position. So this issue may not be fixable.

**Comment by lucasvreis** [comment]
Oh, you've got me thinking about it and it really sounds difficult. Perhaps instead of tracking where forms are with overlays, the template itself could be stored in the fields' overlay's state, along with each field's value. Upon field modification, replace the whole range of the current template expansion according to the original template, but with the field values already supplied. It sounds like this could work with minibuffer prompts without triggering them again, since the previous answers would be already stored in the state. And currently FORM elements are already re-evaluated whenever a field modified, so this wouldn't change.

**Comment by lucasvreis** [comment]
@minad I realized it's possible to do something relatively simple to make sure this kind of error is avoided. It uses the fact the overlays are added to the state in order to make sure they stay in order whenever a field is modified. Check out https://github.com/lucasvreis/tempel/commit/f6f3ec0b77ccfa86922842df75fa6de555d9f1b6.


<img src="https://github.com/minad/tempel/assets/4229115/1e9a7e3a-c029-45a4-a0f7-1e4774f3808e" width="300">

**Comment by lucasvreis** [comment]
I think there was a small "issue" in tempel that was not manifesting before my modification, `tempel--synchronize-fields` attempts to modify the `range` overlay but it is also being called during `tempel--field`, where the `range` overlay does not exist yet (so it was modifying the field overlay itself, which was not a problem anyway). But now I will need to change this to make the modification work in more cases.

**Comment by lucasvreis** [comment]
I think I've solved most of the previous issues: https://github.com/minad/tempel/compare/main...lucasvreis:tempel:main. If you think it makes sense I can open a PR.

**Comment by lucasvreis** [comment]
The snippet in #99 also works after this

![output-2024-06-05-01:24:38](https://github.com/minad/tempel/assets/4229115/9ce7817a-916f-4c27-a861-b87d722ab0fc)

**Comment by minad** [comment]
Great! Please open a PR!
